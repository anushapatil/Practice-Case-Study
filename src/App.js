import React, { useEffect } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux' 
import { useHistory } from 'react-router'

import { Dropdown, Table } from './components'
import {
  GET_ROUTES_REQUESTED,
  GET_DIRECTIONS_REQUESTED,
  GET_STOPS_REQUESTED,
  GET_DEPARTURES_REQUESTED
} from './redux/actions'
import GlobalStyle from './theme/globalStyle'

const App = ({
  routeId,
  routes,
  showDirection,
  directionId,
  directions,
  showStop,
  stopId,
  stops,
  showDeparture,
  getRoutes,
  getDirections,
  getStops,
  getDepartures,
  urlRouteId,
  urlDirectionId,
  urlStopId,
}) => {
  const history = useHistory()
  useEffect(() => {
    getRoutes()
  }, [])
  useEffect(() => {
    if ((urlRouteId && urlDirectionId && urlStopId ) && (urlRouteId != routeId || urlDirectionId != directionId || urlStopId != stopId)) {
      // get departure table
      getDepartures(urlRouteId, urlDirectionId, urlStopId)
    }
  }, [urlRouteId, urlDirectionId, urlStopId])

  const handleGetDepartures = (stopId) => {
    // push id to the url path
    history.push(`/${routeId}/${directionId}/${stopId}`)
    // dispatch action to get directions
    getDepartures(routeId, directionId, stopId)
  }
  return (
    <div className='container'>
      <GlobalStyle />
      <h2 className='text-center'>Real-time Departures</h2>
      <Dropdown 
        list={routes}
        value='route_id'
        description='route_label'
        callback={getDirections}
        defaultId={routeId}
      />
      { showDirection &&
        <Dropdown 
          list={directions}
          value='direction_id'
          description='direction_name'
          callback={getStops}
          defaultId={directionId}
        />
      }
      { showStop &&
        <Dropdown 
          list={stops}
          value='place_code'
          description='description'
          callback={handleGetDepartures}
          defaultId={stopId}
        />
      }
      { showDeparture &&
        <Table />
      }
    </div>
  )
}

App.propTypes = {
  showDirection: PropTypes.bool,
  routes: PropTypes.array,
  showStop: PropTypes.bool,
  directions: PropTypes.array,
  stops: PropTypes.array,
  showDeparture: PropTypes.bool,
  getRoutes: PropTypes.func,
  getDirections: PropTypes.func,
  getStops: PropTypes.func,
  getDepartures: PropTypes.func
}

const mapStateToProps = (state) => ({
  routeId: state.transit.routeId,
  routes: state.transit.routes,
  showDirection: state.transit.showDirection,
  directionId: state.transit.directionId,
  directions: state.transit.directions,
  showStop: state.transit.showStop,
  stopId: state.transit.stopId,
  stops: state.transit.stops,
  showDeparture: state.transit.showDeparture
})

const mapDispatchToProps = (dispatch) => ({
  getRoutes: () => dispatch({ type: GET_ROUTES_REQUESTED }),
  getDirections: (routeId) => dispatch({ type: GET_DIRECTIONS_REQUESTED, routeId }),
  getStops: (directionId) => dispatch({ type: GET_STOPS_REQUESTED, directionId }),
  getDepartures: (routeId, directionId, stopId) => dispatch({ type: GET_DEPARTURES_REQUESTED, routeId, directionId, stopId })
})

export default connect(mapStateToProps, mapDispatchToProps)(App)
