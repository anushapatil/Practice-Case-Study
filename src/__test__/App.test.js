import React from 'react'
import { fireEvent, render, act } from '@testing-library/react'
import configureStore from 'redux-mock-store'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'

import App from '../App'

const mockStore = configureStore([])
jest.mock('react-router', () => ({
	useHistory: () => ({
		push: jest.fn(),
	}),
}))

describe('App Connected React-Redux Component', () => {
	let store

	beforeEach(() => {
		store = mockStore({
			transit: {
				routeId: '901',
				routes: [
					{
						agency_id: 0,
						route_id: '901',
						route_label: 'Select route'
					},{
					agency_id: 1,
					route_id: '902',
					route_label: 'METRO Blue Line'
				}],
				showDirection: true,
				directionId: 1,
				directions: [{
					direction_id: 0,
					direction_name: 'Select direction'
				},{
					direction_id: 1,
					direction_name: 'Eastbound'
				}],
				showStop: true,
				stopId: 'UNDP',
				stops: [{
					description: 'Select stop',
					place_code: 'SELECT_STOP'
				},{
					description: 'Union Depot',
					place_code: 'UNDP'
				}],
				showDeparture: true,
				getRoutes: jest.fn(),
				getDirections: jest.fn(),
				getStops: jest.fn(),
				getDepartures: jest.fn(),
				departures: [{
					actual: true,
					departure_text: '6 Min',
					departure_time: 1607571120,
					description: 'to Mpls-Target Field',
					direction_id: 1,
					direction_text: 'WB',
					route_id: '902',
					route_short_name: 'Green',
					schedule_relationship: 'Scheduled',
					stop_id: 56026,
					trip_id: '17807892-DEC20-RAIL-Weekday-01'
				}], 
				departsDescription: 'MOA station',
  			departsStopId: 1
			}
		})

		store.dispatch = jest.fn()
	})

  it('should render with given state from Redux store', () => {
		const appComponent = renderer.create(
			<Provider store={store}>
				<App />
			</Provider>
		)
		expect(appComponent.toJSON()).toMatchSnapshot()
  })
 
  it('should dispatch an action when selected route from dropdown', () => {
		// Setup
  	const { getByTestId } = render(
			<Provider store={store}>
				<App urlRouteId={'901'} urlDirectionId={1} urlStopId={'UNDP'} />
			</Provider>)

		// Execute
		fireEvent.change(getByTestId('stops_dropdown'), {
			target: { value: 'UNDP' },
		})

		// Assert
		expect(store.dispatch).toHaveBeenCalledTimes(2) // For getRoutes and getDepartures
	})
	
	it('should dispatch an action when browser url changed', () => {
		// Setup
		act(() => {
  		render(
				<Provider store={store}>
					<App urlRouteId={'902'} urlDirectionId={1} urlStopId={'UNDP'} />
				</Provider>
			)
		})
		// Assert
		expect(store.dispatch).toHaveBeenCalledTimes(2) // For getRoutes and getDepartures from useEffect
  })
})