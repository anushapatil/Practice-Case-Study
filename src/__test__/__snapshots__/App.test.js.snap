// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Connected React-Redux Component should render with given state from Redux store 1`] = `
<div
  className="container"
>
  <h2
    className="text-center title"
  >
    Real-time Departures
  </h2>
  <div
    className="justify-content-center row mb-3"
  >
    <select
      className="sc-AykKC iTPQsP col-8 border border-secondary"
      onChange={[Function]}
    >
      <option
        data-testid="select-option"
        selected={true}
        value="901"
      >
        Select route
      </option>
      <option
        data-testid="select-option"
        selected={false}
        value="902"
      >
        METRO Blue Line
      </option>
    </select>
  </div>
  <div
    className="justify-content-center row mb-3"
  >
    <select
      className="sc-AykKC iTPQsP col-8 border border-secondary"
      onChange={[Function]}
    >
      <option
        data-testid="select-option"
        selected={false}
        value={0}
      >
        Select direction
      </option>
      <option
        data-testid="select-option"
        selected={true}
        value={1}
      >
        Eastbound
      </option>
    </select>
  </div>
  <div
    className="justify-content-center row mb-3"
  >
    <select
      className="sc-AykKC iTPQsP col-8 border border-secondary"
      data-testid="stops_dropdown"
      onChange={[Function]}
    >
      <option
        data-testid="select-option"
        selected={false}
        value="SELECT_STOP"
      >
        Select stop
      </option>
      <option
        data-testid="select-option"
        selected={true}
        value="UNDP"
      >
        Union Depot
      </option>
    </select>
  </div>
  <div
    className="sc-AykKD kpILEN pl-0 pr-0 pb-3 bg-light bg-secondary"
  >
    <div
      className="title d-flex p-3"
    >
      <h4>
        MOA station
      </h4>
      <span>
        <strong>
          Stop #:
        </strong>
        1
      </span>
    </div>
    <div
      className="row ml-0 mr-0 bg-warning"
    >
      <div
        className="col-3 p-3 text-left font-weight-bold"
      >
        ROUTE
      </div>
      <div
        className="col-5 p-3 text-left font-weight-bold"
      >
        DESTINATION
      </div>
      <div
        className="col-4 p-3 text-right font-weight-bold"
      >
        DEPARTS
      </div>
    </div>
    <div
      className="row border-bottom ml-0 mr-0"
    >
      <div
        className="col-3 p-3 text-left font-weight-bold"
      >
        Green
      </div>
      <div
        className="col-5 p-3 text-left"
      >
        to Mpls-Target Field
      </div>
      <div
        className="col-4 p-3 text-right font-weight-bold"
      >
        6 Min
      </div>
    </div>
  </div>
</div>
`;
